src/storage/leveldb/utxo_index.ts

```diff
@@ -0,0 +1,220 @@
+    try {
+      const dbVersion = await db.get('version');
+      if (this.indexVersion !== dbVersion) {
+        throw new Error(`Database version mismatch for ${this.constructor.name}: database version (${dbVersion}) expected version (${this.indexVersion})`);
+      }
+    } catch (err: unknown) {
+      if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+        // This is a new db, add version and return
+        await db.put('version', this.indexVersion);
+        return;
+      }
+      throw err;
+    }
+  }
+
+  async validate(): Promise<void> {
+    await this.checkVersion();
+
+    // Iterate on all addresses and check that we have a corresponding index entry
+    for await (const [key, value] of this.utxoDB.iterator()) {
+      if (key !== _utxo_id(value)) {
+        throw new Error('Inconsistent database');
+      }
+
+      try {
+        const tokenUtxo = await this.tokenUtxoDB.get(_token_utxo_key(value));
+        if (!_.isEqual(tokenUtxo, value)) {
+          throw new Error('Inconsistent database')
+        }
+
+      } catch(err: unknown) {
+        if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+          // Create if missing
+          await this.tokenUtxoDB.put(_token_utxo_key(value), value);
+        } else {
+          throw err;
+        }
+      }
+
+      try {
+        const tokenAddrUtxo = await this.tokenAddressUtxoDB.get(_token_address_utxo_key(value));
+        if (!_.isEqual(tokenAddrUtxo, value)) {
+          throw new Error('Inconsistent database')
+        }
+      } catch(err: unknown) {
+        if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+          // Create if missing
+          await this.tokenAddressUtxoDB.put(_token_address_utxo_key(value), value);
+          continue;
+        }
+        throw err;
+      }
+    }
+  }
+
+  async * utxoIter(): AsyncGenerator<IUtxo> {
+    for await (const utxo of this.utxoDB.values()) {
+      yield utxo;
+    }
+  }
+
+  async * selectUtxos(options: IUtxoFilterOptions, networkHeight?: number): AsyncGenerator<IUtxo> {
+    const isHeightLocked = (utxo: IUtxo) => {
+      if (utxo.type !== BLOCK_VERSION) {
+        // Only blocks can be reward locked
+        return false;
+      }
+      if (!(options.reward_lock && networkHeight)) {
+        // We do not have details to process reward lock
+        return false;
+      }
+      // Heighlocked when network height is lower than block height + reward_spend_min_blocks
+      return ((utxo.height || 0) + options.reward_lock) > networkHeight;
+    };
+
+    const token = options.token || HATHOR_TOKEN_CONFIG.uid;
+    const authorities = options.authorities || 0;
+    const maxUtxos = options.max_utxos || MAX_INPUTS;
+
+    let db: typeof this.utxoDB;
+    const itOptions: ValueIteratorOptions<string, IUtxo> = {};
+    if (options.filter_address !== undefined) {
+      // Use tokenAddressUtxoDB
+      db = this.tokenAddressUtxoDB;
+      let minkey = `${authorities}:${token}:${options.filter_address}:`;
```

Comment:
Having the it key ending in `:` is ok?

Reply:
Yes, since iteration is sorted on keys.
You can see below that if we don't add a value filter for the amount (`options.amount_smaller_than`) we will calculate the `maxaddr` which is the `options.filter_address` but with the last char 1 higher than the address we are filtering, this actually gives us all keys with the address as a prefix.